### ISSUE TITLE
Higher validation loss (but same training loss) when using multi_gpu_model

### ISSUE THREAD

**Issue by thorstenwagner** [keras-team/keras#11236]
Hi,

I'm facing an increase in validation loss when using multi_gpu_model to train a CNN. I kept the all other parameters constant (batchsize, learning rate).

This is how I implemented it:

 I created the single-gpu model:
  
      if num_gpus > 1:
        import tensorflow as tf
        with tf.device('/cpu:0'):
            model = create_model()

Then I converted it into a parallel model:

    if num_gpus > 1:
        parallel_model = multi_gpu_model(model, gpus=num_gpus)

Then I compile it and run the training:

    parallel_model.compile(...)
    parallel_model.fit_generator(...)

Any ideas what might be the source of error?

I'm using the latest Keras (current master) with the latest tensorflow (1.10.1).

Best,
Thorsten

**Comment by gabrieldemarmiesse** [comment]
Can you please give us a minimal script to reproduce the error? Thank you.

**Comment by thorstenwagner** [comment]
Unfortunately I cannot reproduce it on a small dataset, only on my big not-yet-public dataset...

**Comment by tedyu** [comment]
Looking at the doc for multi_gpu_model :
```
    - Concatenate the results (on CPU) into one big batch.
```
It seems the concatenation is done at the end.
So I don't quite know what you meant by 'after one step' above.

**Comment by thorstenwagner** [comment]
"One step" is for me one forward pass and backward pass of a batch. 

In the meantime I've updated to python3.6 as I thought the problem might be python2 specific. But it is not.

Could be my batch generator be the problem? I'm using a sequence object for doing this but if it works for a single gpu I cannot see why it should not work for multi gpu.

Any recommendation how to track down the problem is highly appreciated.

**Comment by thorstenwagner** [comment]
It seems that I fixed it. The problem was, that I overwrote at the template_model with the parallel_model before training. I fixed that and now I get comparable loss values.

**Comment by thorstenwagner** [comment]
I"m sorry, but I just mixed up the parameters. I checked it today more carefully again and the problem is still there :-( 

Just to provide you a comparison, I made three runs:
After 5 Epochs on single gpu: 0.0905, 0.086,0.088
After 5 Epochs using multi gpu: 0.0983, 0.11219, 0.10684

Any recommendation how to track down the problem?

**Comment by thorstenwagner** [comment]
Could this issue be reopened?

**Comment by thorstenwagner** [comment]
I made some more experiments. Again I run the training three times for 5 epochs with multi and with single gpu:
![multi_gpu_experiments](https://user-images.githubusercontent.com/1113977/47009068-d9abf100-d13b-11e8-8d04-5a04f9b5fb7a.png)
As you can see, the training loss behaving similar very similar with mutliple and single gpu.
However, the validation loss is always worse.

**Comment by rdiazgar** [comment]
Hi,

I'm facing the same problem here when using a multi_gpu_model. I had to set cpu_merge=False to make the training loss stable after the first epoch. However, the validation loss is too high. 

Did you eventually solve this?

Thanks
Raul

**Comment by thorstenwagner** [comment]
Hi Raul,

Unfortunately not :-(

Best,
Thorsten

**Comment by rdiazgar** [comment]
That's a pity. I will let you know if I figure this out.

Thanks
Raul

**Comment by rdiazgar** [comment]
I torsten,

It seems that, for now, it is working in my case. My set up is the following:

- Keras 2.2.4
- TF 1.12.0
- I am training the keras port of DeepLabv3+ (https://github.com/bonlime/keras-deeplab-v3-plus/)

My model declaration is pretty standard. I don't even declare the model under the scope of the cpu because it did not solve the problem. For me, what it fixed everything is to allow the merge in the GPU, and to create a slightly different model checkpoint callback that saves the single-GPU model instead of the multi-GPU one. So far it works in a 2-GPU setup, and the validation loss reports as expected. I hope this can be of help to you.

Raul

```
    single_model = Deeplabv3(input_shape=crop_size + (3,), classes=len(ranges), backbone='xception')

    # muli-gpu
    if gpus > 1:
        callbacks.append(MultiGPUModelCheckpoint(single_model, weights_path, monitor=monitor, verbose=0, save_best_only=True,
                                                 save_weights_only=True, mode='auto', period=1))
        model = multi_gpu_model(single_model, gpus=gpus, cpu_merge=False)
    else:
        callbacks.append(keras.callbacks.ModelCheckpoint(weights_path, monitor=monitor, verbose=0, save_best_only=True,
                                                         save_weights_only=True, mode='auto', period=1))
        model = single_model

    model.compile(loss=loss, optimizer=optimizer, metrics=metric)
    # model.summary()
    history = model.fit_generator(train_crop_generator, steps_per_epoch=train_steps_per_epoch,
                                  validation_data=val_crop_generator, validation_steps=val_steps_per_epoch,
                                  epochs=epochs, callbacks=callbacks)

```

**Comment by angeloyeo** [comment]
Hi Raul. A few questions.

In your previous post on Jan 9th, you said you failed to get proper val_loss even with cpu_merge=False option. In the post on Jan 10th, which would be right above mine, the cpu_merge option is still False, but you are saying you succeeded to get desired val_loss. 

So the question is ... What was the most important change you applied to get the last result? Is it the callback you used in the code? How can a callback make a different result in terms of val_loss?

Best, 
Donghoon

**Comment by rauldiaz** [comment]
Hi @angeloyeo ,

To this date, I have no certainty why the setup worked. It was more a sort of trial and error thing. The callback definitely helped in being able to re-use the weights elsewhere. That said, a 4-GPU setup of my code above did not work too well either. There certainly seems to be something off with Keras and multi-gpu that needs to be fixed at its core.

**Comment by QtacierP** [comment]
I have met the same problem. I tried to change callback to save weight with singel_model rather than multi-GPU-model, and it works well. I don't understand why I need to change CheckPoint Option, because I just need to verify on my val and test data without loading processing (Just for the compared experiment)
